### PR TITLE
Masked data iterator

### PR DESCRIPTION
**Pull request by Bam4d**
A data iterator that includes a mask vector, can be used in NLP tasks where gradients only need to be calculated for specific time steps.

### CODE REVIEW

**Comment by yinyinl** [comment]
Did you revert commit c16b796 intentionally? Do you plan to do a PR for those in the future?

File: neon/data/dataiterator.py [hunk 1]
```diff
@@ -134,6 +134,135 @@ def __iter__(self):
             targets = self.ybuf if self.ybuf else inputs
             yield (inputs, targets)
 
+class MaskedDataIterator(NervanaObject):
```
Comment (yinyinl) [inline neon/data/dataiterator.py]: Do you think this class can be derived from DataIterator or added into it?

**Comment by yinyinl** [comment]
Thanks for the PR! Have a few questions inline...

**Comment by Bam4d** [comment]
Yes, the tree LSTM is on the backburner for me at the moment, so don't want to commit it just yet.

I will attach a unit in due course

**Comment by Bam4d** [comment]
Closing due to history issues, will re-open

**Pull request closed without merging**